### PR TITLE
Add "Recent" file listing

### PR DESCRIPTION
**Pull request by icewind1991**
Show a list of recently modified files and folders

recent
### On showing folders

How to show folders in the list is the most complex part of this PR, since the mtime (both on disk and in cache) of a folder gets changed when a file in the folder is modified simply just showing folders in the same way as files would result in the parent folder being shown together with the most recently changed file in that folder.

Ideally we would show folders based on their creation time but unfortunately we don't save that information, instead this PR uses the modified date of the least recently changed file in the folder which isn't perfect (especially for folders containing a single file) but it works decently.

I could also be better to hide most folders altogether in most cases
### On hiding folder contents

Another thing I was thinking about to make the recent list more useful is hiding the contents of newly uploaded folders, currently if you upload a large folder the contents of it will take over the entire recents list.
Instead it would be nice if the recents list would only show the newly uploaded folder, with the newly uploaded contents being implicitly grouped in the folder.
### Further work (maybe for nc+1)
- [x] unit tests
- [ ] make the list as useful as possible
  - [ ] maybe hide folders where they don't add value
  - [ ] hide content of folders where they don't add value
- [x] Icon for the files sidebar
- [ ] Maybe allow specifying the range mtimes to be searched for (show me files modified 2 days ago)

cc @jancborchardt @LukasReschke

### CODE REVIEW

**Comment by mar1u5** [comment]
Wow... Amazing 👊 
How is Google Drive, OneDrive or Dropbox handling folder contents?

**Comment by Bugsbane** [comment]
Love, love, love this! I can create an icon if needed. My first thought though, is that a "reload" type icon would work just fine here though. I mean like an arrow going around in a circle.

**Comment by MorrisJobke** [comment]
> What think the rest of @nextcloud/designers? My preference is something Nautilus-like (despite being a KDE fanboy) such as this.

cc @nextcloud/designers

File: apps/files/lib/Controller/ApiController.php [hunk 1]
```diff
@@ -171,6 +180,36 @@ public function getFilesByTag($tagName) {
 	}
 
 	/**
+	 * Returns a list of recently modifed files.
+	 *
+	 * @NoAdminRequired
+	 *
+	 * @return DataResponse
+	 */
+	public function getRecentFiles() {
```
Comment (LukasReschke) [inline apps/files/lib/Controller/ApiController.php]: Unit test for this? 😉

**Comment by williambargent** [comment]
:+1: Looks good, Just need the Icon. @Bugsbane I like the idea of some sort of clock face. Would be awesome if we could get it to actually show the time of the last file which was modified (Would probably cause to many issues).
time-512

File: apps/files/lib/Controller/ApiController.php (hunk 1 above)
Comment (icewind1991) [inline apps/files/lib/Controller/ApiController.php]: In the todo :smile:

**Comment by icewind1991** [comment]
> How is Google Drive, OneDrive or Dropbox handling folder contents?

It gets all it's data from the cache, as long as changes are scanned they should show up

**Comment by jancborchardt** [comment]
Ok, so how does this exactly compare to »Sort by modified date«?

I think the general thing is cool and all, I am just worried we're adding more and more stupid sections to the navigation without properly connecting everything. That is, I think a much better solution would be to:
- use »Sort by modified date« as default in the all files list
- of course remember if that sorting is changed, on a per-user basis
- (just to bring an example of another nav section we can integrate better) we should always sort the favorite files of that folder up top

**Comment by icewind1991** [comment]
> Ok, so how does this exactly compare to »Sort by modified date«?

The recents "tab" shows recent files across all folders, not just the current folder

**Comment by Bugsbane** [comment]
This is exactly the same as how file browsers let you both sorry by date, but also see all of the files you've modified recently, regardless of location. I use both daily, for different purposes.

**Comment by jancborchardt** [comment]
It was partly a rhetorical question. ;) What I meant to say is that my proposal above would already take care of a majority of cases for the »Recent« section, but in a less interface-filling and more integrated way.

Let's take small steps, and find how we can do more with less. Again sure, we can just add stuff, but that's the trap software falls into before people abandon it cause it's too complicated.

**Comment by Bugsbane** [comment]
draft mockup of a clock style icon for recently modified files:
recent-icon-1

**Comment by jancborchardt** [comment]
@Bugsbane very nice! :) I'd only increase the size by 1px for each side because compared to the square icons around it looks a tad small. That's because it has the same width as the folder or share icons, but being a circle it needs a bit more size to appear visually sized correctly.

**Comment by Bugsbane** [comment]
Ok, this is as large as it can go, while still fitting into the 16x16 png image size:
recent-icon-1

If it's ok, just let me know and I'll svgo/trimage it and commit.

**Comment by icewind1991** [comment]
> `png`

master will no longer have fallback png's so no need to add a png version

> what path / filename should I commit the svg/png to?

You'll need to add the `nav-icon-recent` class (see apps/files/css/files.css line ~90) which points to the newly added icon

**Comment by Bugsbane** [comment]
Ah, ok. It's being pulled in by css. Thanks. I just did a quick search on the commit changes page for .png and .svg and didn't see anything, and assumed you hadn't added the icon yet as it didn't exist. I'll add the class.

**Comment by jancborchardt** [comment]
@Bugsbane looks great! Please make sure the circle is pixel-perfect, that is it should be either 1px or most probably 2px width. That will ensure it will look sharp and not blur.

**Comment by Bugsbane** [comment]
Ok, I've submitted PR https://github.com/nextcloud/server/pull/340 and made sure that as much as is possible, that everything uses whole pixel stroke widths and aligns with the pixel grid. One thing puzzles me though. I could see the class pulling in the svg, but I didn't see anything referring to the PNG fallbacks, although I figure that these substitutions are probably done automatically by JS rather than being hard coded.

**Comment by Bugsbane** [comment]
Oh, and I tried svgo on it, but no matter how I tried to modify the original SVG, svgo always butchered the image, completely removing the hands of the clock or worse. Accordingly I ran sourge over it instead, which worked fine.

**Comment by icewind1991** [comment]
@LukasReschke @rullzer @nickvergessen please review

**Comment by jancborchardt** [comment]
@icewind1991 when just switching to the branch, you get a »No files in here« message. Which is weird cause I assumed the »Recent« value is just taken from the modified time. If this list is not populated by default, it will be quite confusing as soon as people update.

**Comment by icewind1991** [comment]
> I assumed the »Recent« value is just taken from the modified time.

It is, can you check if there is anything in the log or check the result of `apps/files/api/v1/recent` call

**Comment by schiessle** [comment]
@icewind1991 Great stuff!

Question: What about the open checkboxes at the top. Are all of them addressed? 

From a first gut feeling I would keep folders out of the list. The file which changed in the folder will be shown anyway. So I don't see the reason to show (all?) parent folders of all recently changes files as well in this view.

**Comment by jancborchardt** [comment]
@icewind1991 the view only seems to get populated once I mess with new files.
How do I check the API call? Just opening the URL gives me ` Access forbidden. CSRF check failed`

And another thing: The »Recent« view also has headers to sort by size and name, and least recently updated. That doesn’t really make sense. At the very least we should reset that when switching back from the view, or ideally completely disable sorting in that view.

**Comment by icewind1991** [comment]
> How do I check the API call?

Open your browser's dev tools, go to the nextwork tab, click on the "Recent" button and see what calls are being made

> And another thing: The »Recent« view also has headers to sort by size and name, and least recently updated. That doesn’t really make sense. At the very least we should reset that when switching back from the view, or ideally completely disable sorting in that view.

True, they are "left overs" from the way the view is being made, will see if I can get rid of it

**Comment by schiessle** [comment]
- [ ] The texteditor no longer loads on this branch, even after rebasing it on master (4a963c21fff251a5ea105e6ec50e6d87dca1654e). Directly on master everything works fine

**Comment by icewind1991** [comment]
- Fixed texteditor
- Disallow sorting
- Fixed no files showing up in some cases

This is good for re-review ( @jancborchardt @schiessle @MorrisJobke ) any open tasks in the original post are things we can look into for a further release

**Comment by MorrisJobke** [comment]
I tested it and it works 👍

**Comment by MorrisJobke** [comment]
```
There was 1 failure:

1) OCA\Files\Tests\Controller\ViewControllerTest::testIndexWithRegularBrowser
Failed asserting that two objects are equal.
--- Expected
+++ Actual
@@ @@
                         'appname' => 'files'
-                        'script' => 'list.php'
+                        'script' => 'recentlist.php'
                         'order' => 2
                         'name' => 'Recent'
                         'active' => false
                         'icon' => ''
                     )
                     2 => Array (...)
                     3 => Array (...)
                     4 => Array (...)
                     5 => Array (...)
                     6 => Array (...)
                     7 => Array (...)
                 )
             )
             'l10n' => OC\L10N\L10N Object (...)
             'theme' => OC_Defaults Object (...)
         )
         'appContents' => Array (...)
     )
     'renderAs' => 'user'
     'appName' => 'files'
     'headers' => Array (...)
     'cookies' => Array ()
     'status' => 200
     'lastModified' => null
     'ETag' => null
     'contentSecurityPolicy' => OCP\AppFramework\Http\ContentSecurityPolicy Object (...)
 )

/drone/src/github.com/nextcloud/server/apps/files/tests/Controller/ViewControllerTest.php:318

```

**Comment by williambargent** [comment]
:+1: tested and works, it is missing the favourites column, I think this should still be included in the view. Without it looks a bit odd transitioning from Files to Recent as everything shifts over to the left.

**Comment by MorrisJobke** [comment]
> Without it looks a bit odd transitioning from Files to Recent as everything shifts over to the left.

The other views also don't have the favorite column. I would leave it like it is now.

**Comment by williambargent** [comment]
For consistency, it might be worth reviewing and adding favourites to all the views.

**Comment by mar1u5** [comment]
Is this something we want to have for Nextcloud 10? @karlitschek

**Comment by MorrisJobke** [comment]
> Is this something we want to have for Nextcloud 10? @karlitschek

I would say no, because this is clearly a feature and we only back port bug fixes.

**Comment by karlitschek** [comment]
Nice. I agree that we shouldn't backport this for now

**Comment by jancborchardt** [comment]
@williambargent can you open a new issue about adding the favorites column to all views? It should indeed be there for everything, including search results.

**Pull request merged**